README
 Description
-------------

The CVSAnalY tool extracts information out of source code repository logs and stores it into a database.
This version is a fork of the original code for ALERT project.
The original project can be found in http://metricsgrimoire.github.com/CVSAnalY/


 License
---------

Licensed under GNU General Public License (GPL), version 2 or later.


 Download
----------

 * git://github.com/AlertProject/CVSAnalY.git
 * https://github.com/AlertProject/CVSAnalY.git


 Requirements
--------------

CVSAnalY has the following dependencies:

    * RepositoryHandler
        git clone git://github.com/MetricsGrimoire/RepositoryHandler.git
    * cvs (optional, for CVS support)
    * subversion (optional, for SVN support)
    * git (optional, for Git support)
    * Python MySQLDB (optional but recommended)
    * Python SQLite (optional)


 Installation
--------------

You can install cvsanaly2 just by running setup.py script:

    # python setup.py install

This will install it in the python default directories in your system.

If you don't install CVSAnalY with root privileges, or don't want
to install it in the default directories, you can also use the source
code directory, as cloned from the main git repo. It is enough to
configure your PATH and PYTHONPATH so that cvsanaly2, and the Python
modules it needs, are found.

Add to your $PATH the directory which contains cvsanaly
(cvsanalydir is the directory where CVSAnalY is installed):

    $ export PATH=$PATH:cvsanalydir

In PYHTONPATH, you need to include both the dirs for CVSAnalY and
RepositoryHandler (except if it is installed in the default Python dirs
in your system). If repohandlerdir is the path where RepositoryHandler
is installed:

    $ export PYTHONPATH=$PYTHONPATH:cvsanalydir:repohandlerdir

You are ready to use cvsanaly2!


 Running cvsanaly2
-------------------

For the impatients: just checkout (from svn or cvs) to obtain a local
version of your repository, and then run cvsanaly2:

    $ cd project/
    ~/project$ cvsanaly2

More options, and a more detailed info about the options, can be
learnt by running "cvsanaly2 --help"

 Improving CVSAnalY
--------------------

Source code, wiki and ITS available on Github:
* https://github.com/MetricsGrimoire/CVSAnalY

If you want to receive updates about new versions, and keep in touch
with the development team, consider subscribing to the mailing list.
It is a very low traffic list (< 1 msg a day):

* https://lists.libresoft.es/listinfo/metrics-grimoire


 Credits
---------

CVSAnalY was initially developed by the GSyC/LibreSoft group at the
Universidad Rey Juan Carlos, Madrid (Spain). It is
part of a wider research on libre software engineering, aimed to gain
knowledge on how libre software is developed and maintained.

For more information about ALERT project please visit http://www.alert-project.eu


 Main authors
--------------

Carlos Garcia Campos <carlosgc at gsyc.es>

 Contributors
--------------

Gregorio Robles <grex at gsyc.escet.urjc.es>
Alvaro Navarro <anavarro at gsyc.escet.urjc.es>
Jesus M. Gonzalez-Barahona <jgb at gsyc.escet.urjc.es>
Israel Herraiz <herraiz at gsyc.escet.urjc.es>
Juan Jose Amor <jjamor at gsyc.escet.urjc.es>
Martin Michlmayr <tbm at debian.org>
Alvaro del Castillo <acs at barrapunto.com>
Santiago Dueñas <sduenas at libresoft.es>


debian/changelog
cvsanaly (2.1.3-1alert) unstable; urgency=low

  * Dependencies updated.
  * Fixed minor bugs in extensions

 -- Santiago Dueñas <sduenas@libresoft.es>  Tue, 30 Apr 2013 17:49:38 +0200

cvsanaly (2.1.2-2alert) unstable; urgency=low

  * Dependencies updated.

 -- Roberto Andradas <randradas@libresoft.es>  Thu, 14 Feb 2013 20:02:46 +0100

cvsanaly (2.1.2-1alert) unstable; urgency=low

  * Debian package for Alert project

 -- Roberto Andradas <randradas@libresoft.es>  Mon, 11 Feb 2013 17:11:21 +0100

cvsanaly (2.1.0-1) unstable; urgency=low

  * New upstream bugfix release:
    + debian/control:
      - Maintainer changed
      - Version number added to the repositoryhandler dependency
    + debian/copyright:
      - Upstream-Contact changed
      - Initial copyright year fixed

 -- Luis Cañas Díaz <lcanas@libresoft.es>  Wed, 28 Oct 2011 18:19:59 +0200

cvsanaly (2.0.0) unstable; urgency=low

  * New package version (Closes: #123456)

 -- Juan Francisco Gato Luis <jfcogato@libresoft.es>  Tue, 12 Jul 2011 13:08:01 +0200

cvsanaly (2.0.0) stable; urgency=low

    * new package version

 -- Juan Francisco Gato <jfcogato@libresoft.es>  Fri, 18 Sep 2009 12:38:00 +2000

cvsanaly (1.0.1) unstable; urgency=low

    * new package

 -- Alvaro Navarro <anavarro@gsyc.escet.urjc.es>  Tue, 11 Apr 2006 19:30:37 +2000



pycvsanaly2/extensions/FileTypes.py
# Copyright (C) 2008 LibreSoft
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
# Authors :
#       Carlos Garcia Campos <carlosgc@gsyc.escet.urjc.es>

from pycvsanaly2.Database import SqliteDatabase, MysqlDatabase, TableAlreadyExists, statement
from pycvsanaly2.extensions import Extension, register_extension, ExtensionRunError
from pycvsanaly2.extensions.file_types import guess_file_type
from pycvsanaly2.utils import to_utf8, uri_to_filename

class DBFileType:

    id_counter = 1

    __insert__ = "INSERT INTO file_types (id, file_id, type) values (?, ?, ?)"

    def __init__ (self, id, type, file_id):
        if id is None:
            self.id = DBFileType.id_counter
            DBFileType.id_counter += 1
        else:
            self.id = id

        self.type = to_utf8 (type)
        self.file_id = file_id

class FileTypes (Extension):
    
    def __init__ (self):
        self.db = None
    
    def __create_table (self, cnn):
        cursor = cnn.cursor ()

        if isinstance (self.db, SqliteDatabase):
            import pysqlite2.dbapi2
            
            try:
                cursor.execute ("CREATE TABLE file_types (" +
                                "id integer primary key," +
                                "file_id integer," +
                                "type varchar" +
                                ")")
            except pysqlite2.dbapi2.OperationalError:
                cursor.close ()
                raise TableAlreadyExists
            except:
                raise
        elif isinstance (self.db, MysqlDatabase):
            import _mysql_exceptions

            try:
                cursor.execute ("CREATE TABLE file_types (" +
                                "id INT primary key," +
                                "file_id integer," +
                                "type mediumtext," +
                                "FOREIGN KEY (file_id) REFERENCES files(id)" +
                                ") ENGINE=MyISAM" +
                                " CHARACTER SET=utf8")
            except _mysql_exceptions.OperationalError, e:
                if e.args[0] == 1050:
                    cursor.close ()
                    raise TableAlreadyExists
                raise
            except:
                raise
            
        cnn.commit ()
        cursor.close ()

    def __get_files_for_repository (self, repo_id, cursor):
        query = "SELECT ft.file_id from file_types ft, files f " + \
                "WHERE f.id = ft.file_id and f.repository_id = ?"
        cursor.execute (statement (query, self.db.place_holder), (repo_id,))
        files = [res[0] for res in cursor.fetchall ()]

        return files
        
    def run (self, repo, uri, db):
        self.db = db

        path = uri_to_filename (uri)
        if path is not None:
            repo_uri = repo.get_uri_for_path (path)
        else:
            repo_uri = uri
        
        cnn = self.db.connect ()

        cursor = cnn.cursor ()
        cursor.execute (statement ("SELECT id from repositories where uri = ?", db.place_holder), (repo_uri,))
        repo_id = cursor.fetchone ()[0]
        
        files = []
        
        try:
            self.__create_table (cnn)
        except TableAlreadyExists:
            cursor.execute (statement ("SELECT max(id) from file_types", db.place_holder))
            id = cursor.fetchone ()[0]
            if id is not None:
                DBFileType.id_counter = id + 1

            files = self.__get_files_for_repository (repo_id, cursor)
        except Exception, e:
            raise ExtensionRunError (str (e))

        query = "select a.file_id fid, f.file_name fname " + \
                "from action_files a, files f " + \
                "where f.id = a.file_id and " + \
                "not exists (select id from file_links where parent_id = a.file_id) " + \
                "and f.repository_id = ? group by fid, fname"

        cursor.execute (statement (query, db.place_holder), (repo_id,))
        write_cursor = cnn.cursor ()
        rs = cursor.fetchmany ()
        while rs:
            types = []

            for file_id, file_name in rs:
                if file_id in files:
                    continue
                
                type = guess_file_type (file_name)
                types.append (DBFileType (None, type, file_id))

            if types:
                file_types = [(type.id, type.file_id, type.type) for type in types]
                write_cursor.executemany (statement (DBFileType.__insert__, self.db.place_holder), file_types)

            rs = cursor.fetchmany ()
            
        cnn.commit ()
        write_cursor.close ()
        cursor.close ()
        cnn.close ()
        

register_extension ("FileTypes", FileTypes)


pycvsanaly2/extensions/Jobs.py
# Copyright (C) 2009 LibreSoft
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
# Authors:
#       Carlos Garcia Campos <carlosgc@libresoft.es>

if __name__ == '__main__':
    import sys
    sys.path.insert (0, "../")
    
from pycvsanaly2.AsyncQueue import AsyncQueue, TimeOut
import repositoryhandler.backends as rh
import threading

class JobPool:

    POOL_SIZE = 5

    def __init__ (self, repo, repo_uri, jobs_done = True, poolsize = POOL_SIZE, queuesize = None):
        self.jobs_done = jobs_done

        self.queue = AsyncQueue (queuesize or 0)
        if self.jobs_done:
            self.done = AsyncQueue ()
            
        for i in range (poolsize):
            rep = repo.copy ()
            thread = threading.Thread (target=self._job_thread, args=(rep, repo_uri))
            thread.setDaemon (True)
            thread.start ()

    def _job_thread (self, repo, repo_uri):
        while True:
            job = self.queue.get ()
            job.run (repo, repo_uri)
            self.queue.done ()

            if self.jobs_done:
                self.done.put (job)

    def push (self, job):
        self.queue.put (job)

    def get_next_done (self, timeout = None):
        if not self.jobs_done:
            return None
        
        try:
            job = self.done.get (timeout)
            self.done.done ()
            return job
        except TimeOut:
            return None

    def get_next_done_unlocked (self):
        if not self.jobs_done:
            return None

        if self.done.empty_unlocked ():
            return None
        
        return self.done.get_unlocked ()

    def join (self):
        self.queue.join ()

class Job:

    def run (self, repo, repo_uri):
        raise NotImplementedError
        

if __name__ == '__main__':
    class JobLastRev (Job):
        def __init__ (self, module):
            self.module = module
            
        def run (self, repo, repo_uri):
            uri = repo_uri + self.module
            print "%s -> %s" % (uri, repo.get_last_revision (uri))

    modules = ['cvsanaly', 'octopus', 'cmetrics', 'repositoryhandler', 'retrieval_system',
               'bicho', 'pandaRest']
    repo = rh.create_repository ('svn', 'https://svn.forge.morfeo-project.org/svn/libresoft-tools/')
    repo_uri = 'https://svn.forge.morfeo-project.org/svn/libresoft-tools/'
    
    pool = JobPool (repo, repo_uri, False)
    for module in modules:
        job = JobLastRev (module)
        pool.push (job)

    pool.join ()
    
